verification/io/io-spec.gobra

```diff
@@ -38,75 +38,221 @@ type Place int
+					v.IO_Internal_val1_1.RightSeg}),
+			v.IO_Internal_val1_3,
+			v.IO_Internal_val1_4)
+}
+
+pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place) {
+	forall v IO_val ::
+	(v.isIO_Internal_val1 ==>
+		(dp3s_iospec_bio3s_enter_guard(s, t, v) ==>
+			(CBio_IN_bio3s_enter(t, v) &&
+				dp3s_iospec_ordered(
+					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
+					CBio_IN_bio3s_enter_T(s, t, v)))))
+}
+
+pred CBio_IN_bio3s_xover_up2down(t Place, v IO_val)
+
+ghost
+requires CBio_IN_bio3s_xover_up2down(t, v)
+pure func dp3s_iospec_bio3s_xover_up2down_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+
+// nextseg.Future[0] == hf2
+// currseg.Future == seq[IO_HF]{hf1}
+
+// hf1.InIF2 === some(recvif)
+// hf1.EgIF2 === some(recvif)
+
+// This corresponds to the condition of the if statement in the io-spec case for xover_up2down
+ghost
+decreases
+pure func dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return let currseg := v.IO_Internal_val1_1.CurrSeg in
+	v.IO_Internal_val1_1.LeftSeg == none[IO_seg2] ? false :
+	let nextseg := get(v.IO_Internal_val1_1.LeftSeg) in
+	(nextseg.ConsDir &&
+		len(nextseg.Future) > 0 &&
+		len(currseg.Future) > 0 &&
+		let hf1, hf2 := currseg.Future[0], nextseg.Future[0] in
+		let traversedseg := dp3s_iospec_bio3s_create_guard_traversedseg(currseg) in
+		(xover_up2down2_link_type(asid(), hf1, hf2) &&
+			dp3s_xover_common(
+				s,
+				v.IO_Internal_val1_1,
+				currseg,
+				nextseg,
+				traversedseg,
+				IO_pkt2(IO_Packet2{nextseg, v.IO_Internal_val1_1.MidSeg, v.IO_Internal_val1_1.RightSeg, some(traversedseg)}),
+				hf1,
+				hf2,
+				v.IO_Internal_val1_2,
+				v.IO_Internal_val1_3,
+				v.IO_Internal_val1_4,)))
+}
+
+pred dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place) {
+	forall v IO_val ::
+	(v.isIO_Internal_val1 ==>
+		(dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) ==>
+			(CBio_IN_bio3s_xover_up2down(t, v) &&
+				dp3s_iospec_ordered(
+					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
+					dp3s_iospec_bio3s_xover_up2down_T(s, t, v)))))
+}
+
+pred CBio_IN_bio3s_xover_core(t Place, v IO_val)
+
+ghost
+requires CBio_IN_bio3s_xover_core(t, v)
+pure func dp3s_iospec_bio3s_xover_core_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+
+// This corresponds to the condition of the if statement in the io-spec case for xover_core
 ghost
-requires CBioIO_bio3s_send(t)
-pure func dp3s_iospec_bio3s_send_R(s IO_dp3s_state_local, t Place) IO_val
+decreases
+pure func dp3s_iospec_bio3s_xover_core_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return (asid() in core_as_set() &&
+		let currseg := v.IO_Internal_val1_1.CurrSeg in
+		(v.IO_Internal_val1_1.LeftSeg == none[IO_seg2] ? false :
+			let nextseg := get(v.IO_Internal_val1_1.LeftSeg) in
+			currseg.ConsDir == nextseg.ConsDir &&
+				len(nextseg.Future) > 0 &&
+				len(currseg.Future) > 0 &&
+				let hf1, hf2 := currseg.Future[0], nextseg.Future[0] in
```

Comment:
why are we indenting this one step? This is specially unclear because the body of the let that declares `hf1` and `hf2` is indented one step less